File: Museo Java/ProcesoRestauracionAutomatica.java
import java.time.LocalDate;
import java.time.temporal.ChronoUnit;
import java.util.ArrayList;
import java.util.Comparator;
import java.util.List;

public class ProcesoRestauracionAutomatica {

    public static final int ANOS_PARA_RESTAURACION = 5;

    private List<Restauracion> restauraciones;

    public ProcesoRestauracionAutomatica(List<Restauracion> restauraciones) {
        this.restauraciones = (restauraciones != null) ? restauraciones : new ArrayList<>();
    }

    public ProcesoRestauracionAutomatica() {
        this(null);
    }

    public List<ObraDeArte> verificarObrasParaRestauracion(List<ObraDeArte> obras) {
        LocalDate hoy = LocalDate.now();
        List<ObraDeArte> obrasARestaurar = new ArrayList<>();

        for (ObraDeArte obra : obras) {
            if (!"expuesta".equals(obra.getEstado())) continue;

            LocalDate ultimaFecha = obtenerFechaUltima(obra);
            LocalDate fechaRef = (ultimaFecha != null) ? ultimaFecha : obra.getFechaEntrada();
            long dias = ChronoUnit.DAYS.between(fechaRef, hoy);
            double anos = dias / 365.25;

            if (anos >= ANOS_PARA_RESTAURACION) {
                obrasARestaurar.add(obra);
                System.out.printf("Obra '%s' requiere restauración (%.1f años desde última intervención).%n",
                        obra.getTitulo(), anos);
            }
        }

        if (obrasARestaurar.isEmpty()) {
            System.out.println("No hay obras que requieran restauración automática hoy.");
        }

        return obrasARestaurar;
    }

    public void actualizarEstadoObras(List<ObraDeArte> obrasARestaurar) {
        for (ObraDeArte obra : obrasARestaurar) {
            obra.setEstado("en_restauracion");
            System.out.println("Obra '" + obra.getTitulo() + "' enviada a restauración automática.");
        }
    }

    private LocalDate obtenerFechaUltima(ObraDeArte obra) {
        return restauraciones.stream()
                .filter(r -> r.getObra().getIdObra() == obra.getIdObra()
                          && "finalizada".equals(r.getEstado()))
                .max(Comparator.comparing(Restauracion::getFechaFin))
                .map(Restauracion::getFechaFin)
                .orElse(null);
    }
}

File: MuseoPOO/proceso_restauracion_automatica.py
"""Módulo que define el proceso automático de restauración."""

from __future__ import annotations

from datetime import date
from typing import TYPE_CHECKING, Optional

if TYPE_CHECKING:
    from obra_de_arte import ObraDeArte
    from restauracion import Restauracion


class ProcesoRestauracionAutomatica:
    """Proceso diario que verifica qué obras necesitan restauración.

    Las obras se restauran automáticamente cada cinco años desde
    su última restauración o fecha de entrada.
    """

    ANOS_PARA_RESTAURACION = 5

    def __init__(
        self,
        restauraciones: Optional[list[Restauracion]] = None,
    ) -> None:
        self._restauraciones = restauraciones or []

    def verificar_obras_para_restauracion(
        self, obras: list[ObraDeArte],
    ) -> list[ObraDeArte]:
        """Verifica qué obras requieren restauración automática."""
        hoy = date.today()
        obras_a_restaurar: list[ObraDeArte] = []

        for obra in obras:
            if obra.get_estado() != "expuesta":
                continue

            ultima_fecha = self._obtener_fecha_ultima(obra)
            fecha_ref = ultima_fecha or obra.get_fecha_entrada()
            dias = (hoy - fecha_ref).days
            anos = dias / 365.25

            if anos >= self.ANOS_PARA_RESTAURACION:
                obras_a_restaurar.append(obra)
                print(
                    f"Obra '{obra.get_titulo()}' requiere "
                    f"restauración ({anos:.1f} años desde "
                    f"última intervención)."
                )

        if not obras_a_restaurar:
            print(
                "No hay obras que requieran restauración "
                "automática hoy."
            )

        return obras_a_restaurar

    def actualizar_estado_obras(
        self, obras_a_restaurar: list[ObraDeArte],
    ) -> None:
        """Actualiza el estado de las obras que necesitan restauración."""
        for obra in obras_a_restaurar:
            obra.set_estado("en_restauracion")
            print(
                f"Obra '{obra.get_titulo()}' enviada a "
                f"restauración automática."
            )

    def _obtener_fecha_ultima(
        self, obra: ObraDeArte,
    ) -> Optional[date]:
        """Obtiene la fecha de la última restauración finalizada."""
        restauraciones_obra = [
            r for r in self._restauraciones
            if r.get_obra().get_id_obra() == obra.get_id_obra()
            and r.get_estado() == "finalizada"
        ]
        if not restauraciones_obra:
            return None
        ultima = max(
            restauraciones_obra,
            key=lambda r: r.get_fecha_fin(),
        )
        return ultima.get_fecha_fin()
